src/java/com/puppetlabs/http/client/impl/ExceptionInsertingPipedInputStream.java

```diff
@@ -0,0 +1,43 @@
+package com.puppetlabs.http.client.impl;
+
+import java.io.IOException;
+import java.io.PipedInputStream;
+
+public class ExceptionInsertingPipedInputStream extends PipedInputStream {
+
+    private final Promise<IOException> ioExceptionPromise;
+
+    public ExceptionInsertingPipedInputStream(Promise<IOException> ioExceptionPromise) {
+        this.ioExceptionPromise = ioExceptionPromise;
+    }
+
+    private void checkFinalResult() throws IOException {
+        try {
+            IOException ioException = ioExceptionPromise.deref();
+            if (ioException != null) {
+                throw ioException;
+            }
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public synchronized int read() throws IOException {
+        int read = super.read();
+        if (read == 1) {
```

Comment:
Yes! The tests didn't catch this as we never actually need this method but I didn't want to leave it not-overidden.